zerver/transaction_tests/test_user_groups.py

```diff
@@ -0,0 +1,155 @@
+                self.subgroup_ids = subgroup_ids
+                self.supergroup_id = supergroup_id
+
+            def run(self) -> None:
+                try:
+                    self.response = test_case.client_post(
+                        url=f"/testing/user_groups/{self.supergroup_id}/subgroups",
+                        info={"add": orjson.dumps(self.subgroup_ids).decode()},
+                    )
+                finally:
+                    # Close all thread-local database connections
+                    connections.close_all()
+
+        def assert_thread_success_count(
+            t1: RacingThread,
+            t2: RacingThread,
+            *,
+            success_count: int,
+            error_messsage: str = "",
+        ) -> None:
+            help_msg = """We access the test endpoint that wraps around the
+real subgroup update endpoint by sychronizing them after the acquisition of the
+first lock in the critical region. Though unlikely, this test might fail as we
+have no control over the scheduler when the barrier timeouts.
+""".strip()
+            barrier = threading.Barrier(parties=2, timeout=3)
+
+            user_group_view.set_sync_after_recursive_query(barrier)
+            t1.start()
+            t2.start()
+
+            succeeded = 0
+            for t in [t1, t2]:
+                t.join()
+                response = t.response
+                if response is not None and response.status_code == 200:
+                    succeeded += 1
+                    continue
+
+                assert response is not None
+                self.assert_json_error(response, error_messsage)
+            # Race condition resolution should only allow one thread to succeed
+            self.assertEqual(
+                succeeded,
+                success_count,
+                f"Exactly {success_count} thread(s) should succeed.\n{help_msg}",
+            )
+
+        foo_chain = self.create_user_group_chain(realm)
+        bar_chain = self.create_user_group_chain(realm)
+        # These two threads are conflicting because a cycle would be formed if
+        # both of them succeed. There is a deadlock in such circular dependency.
+        assert_thread_success_count(
+            RacingThread(
+                subgroup_ids=[foo_chain[0].id],
+                supergroup_id=bar_chain[-1].id,
+            ),
+            RacingThread(
+                subgroup_ids=[bar_chain[-1].id],
+                supergroup_id=foo_chain[0].id,
+            ),
+            success_count=1,
+            error_messsage="Deadlock detected",
+        )
+
+        foo_chain = self.create_user_group_chain(realm)
+        bar_chain = self.create_user_group_chain(realm)
+        # Both threads will attempt to grab a lock on overlapping rows when
+        # they first do the recursive query for subgroups. In this case, we
+        # expect that one of the threads fails due to nowait=True for the
+        # .select_for_update() call.
```

Comment:
Maybe extend this to say start with "These two requests would succeed if they didn't race with each other. However, both ..."